text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+```
+
+This RFC proposes adding two element modifiers: `{{did-render}}` and
+`{{will-destroy}}`.
+
+#### `{{did-render}}`
+
+This modifier runs immediately after the element is inserted, and again whenever
+any of the arguments passed to it update, including the function passed as the
+first argument. It has the following timing semantics:
+
+* **Always**
+  * called _after_ the enclosing component's `didUpdateArgs` hook
+  * called _after_ any child element's `{{did-render}}` modifiers
+  * called in definition order in the template
+* **May or May Not**
+  * be called in the same tick as DOM insertion
+  * have the sibling nodes fully initialized in DOM
+
+> Note: These timing semantics were mostly defined in the
+> [element modifier manager RFC](https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md)
+> and are repeated here for clarity.
+
+`{{did-render}}` receives a function with the following signature as the first
+positional parameter:
+
+```ts
+type DidRenderHandler = (element: Element, ...args): void;
+```
+
+The `element` argument is the element that the modifier is applied to, and the
+rest of the arguments are any remaining positional parameters passed to
+`{{did-render}}`. If the first positional parameter is not a callable function,
+`{{did-render}}` will throw an error. Functions which are passed to this element
+modifier will _not_ be bound to any context by default. Users can bind them
+using the `(action)` helper:
+
+```hbs
+<div {{did-render (action this.setupElement)}}>
+```
+
+#### `{{will-destroy}}`
+
+This modifier runs immediately before the element is removed. It has the
+following timing semantics:
+
+* **Always**
+  * called _after_ the enclosing component's `didUpdateArgs` hook
+  * called _before_ the enclosing component's `willDestroy` hook
+  * called _after_ any child element's `{{will-destroy}}` modifiers
+  * called in definition order in the template
+* **May or May Not**
+  * be called in the same tick as DOM insertion
+  * have the sibling nodes fully initialized in DOM
+
+> Note: These timing semantics were mostly defined in the
+> [element modifier manager RFC](https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md)
+> and are repeated here for clarity.
+
+`{{will-destroy}}` receives a function with the following signature as the first
+positional parameter:
+
+```ts
+type WillDestroyHandler = function(element: Element, ...args): void;
+```
+
+The `element` argument is the element that the modifier is applied to, and the
+rest of the arguments are any remaining positional parameters passed to
+`{{will-destroy}}`. If the first positional parameter is not a callable function,
+`{{will-destroy}}` will throw an error. Functions which are passed to this element
+modifier will _not_ be bound to any context by default. Users can bind them
+using the `(action)` helper:
+
+```hbs
+<div {{will-destroy (action this.teardownElement)}}>
+```
```

Comment:
I think it would be worth having some examples, either here or in the "How we teach this" section, that cover how you'd model common uses of `didInsertElement` with the new element modifiers. I think there are some particularly cool examples that highlight the benefit of this approach, like conditionals (as already mentioned here) as well as wanting to apply some DOM mutation to multiple items in a list.

Reply:
In particular, I'm very interested in plans for bridging to a mental model where elements have their own lifecycles, and the interplay with the component's lifecycle around them

i.e., if you have a set of four tabs, each with a D3 visualization, do you really want things torn down and initialized as you go from tab-to-tab? Maybe we need to have some increased emphasis on making good decisions around use of `{{#if`, `visibility: none`, etc... and the ramifications on element lifecycles